src/corelibs/U2View/src/ov_sequence/view_rendering/SequenceViewAnnotatedRenderer.cpp

```diff
@@ -188,16 +187,56 @@ void SequenceViewAnnotatedRenderer::drawAnnotation(QPainter& p, const QSize& can
                     const QString aText = prepareAnnotationText(aData, as);
                     drawBoundedText(p, annotationRect, aText);
                 }
-                if (simple &&
-                    annotationRect.width() > MIN_WIDTH_TO_DRAW_EXTRA_FEATURES &&
-                    displaySettings.displayCutSites) {
-                    drawCutSite(p, aData, r, annotationRect, as->color, canvasSize, visibleRange);
-                }
             }
             drawAnnotationConnections(p, a, as, displaySettings, canvasSize, visibleRange);
-        } else if (isRestrictionSite && simple && displaySettings.displayCutSites) {
-            p.setPen(borderPen);
-            drawCutSite(p, aData, r, QRect(), as->color, canvasSize, visibleRange);
+        }
+    }
+    // draw restriction sites
+    /**
+     * Calculates restriction cuts positions depends on visible area and calls drawCutSite.
+     *
+     * \param reg Region of the annotation, which sites should be drawn.
+     * \param drawDirect Draw direct cut if true and do not draw if false.
+     * \param drawComplement Draw complement cut if true do not draw if false.
+     */
+    auto drawCutSiteLambda = [this, &p, &visibleRange, &a, &as, &availableHeight, &canvasSize, selected, &aData](const U2Region& reg, bool drawDirect = true, bool drawComplement = true) {
+        if (reg.intersects(visibleRange)) {
+
+            const U2Region visibleLocation = reg.intersect(visibleRange);
+            const U2Region y = getAnnotationYRange(a, 0, as, availableHeight);
+            CHECK(y.startPos >= 0, );
+
+            const int x1 = posToXCoord(visibleLocation.startPos, canvasSize, visibleRange);
+            const int x2 = posToXCoord(visibleLocation.endPos(), canvasSize, visibleRange);
+
+            const int rw = qMax(selected ? MIN_SELECTED_ANNOTATION_WIDTH : MIN_ANNOTATION_WIDTH, x2 - x1);
+            SAFE_POINT(rw > 0, "Negative length of annotationYRange", );
+
+            const QRect annotationRect(x1, y.startPos, rw, y.length);
+            if (annotationRect.width() > MIN_WIDTH_TO_DRAW_EXTRA_FEATURES) {
+                drawCutSite(p, aData, reg, annotationRect, as->color, canvasSize, visibleRange, drawDirect, drawComplement);
+            }
+        } else {
+            drawCutSite(p, aData, reg, QRect(), as->color, canvasSize, visibleRange);
+        }
+    };
+
+    if (isRestrictionSite && displaySettings.displayCutSites) {
+        p.setPen(borderPen);
+        const bool isCircular = ctx->getSequenceObject()->isCircular();
+        auto seqLength = ctx->getSequenceLength();
+        auto merged = U1AnnotationUtils::mergeAnnotatedRegionsAroundJunctionPoint(location, seqLength);
+        if (isCircular && U1AnnotationUtils::isAnnotationContainsJunctionPoint(merged)) {
+            SAFE_POINT(merged.size() == 1, "Unexpected merged enzyme region", );
+
+            auto regionPair = merged.front();
+            drawCutSiteLambda(regionPair.first, true, false);
```

Comment:
Add a comment what is going on here.

Reply:
OK

Reply:
The comment does not answer the question why a circular sequence with a junction point needs to call drawCutSiteLambda twice, separately for each strand.

Look: the `merged` is a concatenation of 2 regions of a circular sequence: (P1...LEN) & (0...P2)
These both regions are located on the same strand.

Why do you draw cut-site for (P1...LEN) on the direct strand only and a cut-site for (0...P2) on the complement?

Reply:
Because we have two regions - (P1...LEN) and (0...P2). We need to calculate `annotationRect` and call `SequenceViewAnnotatedRenderer::drawCutSite` for each region separately.

Reply:
Why the first region is predefined to be `bool drawDirect = true` and the second `bool drawComplement = true` ?

Reply:
It does not have some hidden meaning. As far as I remember, first `drawCutSiteLambda` had more calls than now (now it's called just three times, but in first version it was 5 times or something like this) and the most part of these calls had `..., true, true)` in the end, so I added these default values. Now it does not make ant sense, so it would be better just to remove default values.